devito/mpi/routines.py

```diff
@@ -1232,6 +1347,109 @@ def _arg_apply(self, *args, **kwargs):
         self._C_memfree()
 
 
+class MPIMsg(MPIMsgBase):
+
+    def __init__(self, name, target, halos):
+        self._target = target
+        self._halos = halos
+
+        super().__init__(name, 'msg', self.fields)
+
+        # Required for buffer allocation/deallocation before/after jumping/returning
+        # to/from C-land
+        self._allocator = None
+        self._memfree_args = []
+
+    def _arg_defaults(self, allocator, alias, args=None):
+        # Lazy initialization if `allocator` is necessary as the `allocator`
+        # type isn't really known until an Operator is constructed
+        self._allocator = allocator
+
+        f = alias or self.target.c0
+        for i, halo in enumerate(self.halos):
+            entry = self.value[i]
+
+            # Buffer shape for this peer
+            shape = []
+            for dim, side in zip(*halo):
+                try:
+                    shape.append(getattr(f._size_owned[dim], side.name))
+                except AttributeError:
+                    assert side is CENTER
+                    shape.append(self._as_number(f._size_domain[dim], args))
+
+            # Allocate the send/recv buffers
+            self._allocate_buffers(f, shape, entry)
+
+        return {self.name: self.value}
+
+
+class MPIMsgBasic2(MPIMsgBase):
+
+    def __init__(self, name, target, halos, fixed=None):
+        self._target = target
+        self._halos = halos
+
+        super().__init__(name, 'msg', self.fields)
+
+        # Required for buffer allocation/deallocation before/after jumping/returning
+        # to/from C-land
+        self._fixed = fixed
+        self._allocator = None
+        self._memfree_args = []
+
+    def _arg_defaults(self, allocator, alias, args=None):
+        # Lazy initialization if `allocator` is necessary as the `allocator`
+        # type isn't really known until an Operator is constructed
+        self._allocator = allocator
+
+        f = alias or self.target.c0
+
+        fixed = self._fixed
+
+        # Build a mapper `(dim, side, region) -> (size)` for `f`.
+        mapper = {}
+        for d0, side, region in product(f.dimensions, (LEFT, RIGHT), (OWNED, HALO)):
+            if d0 in fixed:
+                continue
+            sizes = []
+            for d1 in f.dimensions:
+                if d1 in fixed:
+                    continue
+                else:
```

Comment:
since you have a `continue` above there's no point in using an `else` here. Deindent